### DISCUSSION TITLE
how to mesh it and save obj without gui?

### DISCUSSION THREAD

**Discussion by Evilom** [NVlabs/instant-ngp#223] [Q&A]
when i generated mesh in windows platform,the result mesh is very bad,and i saw a youtuber says linuxs can fix this problam, so i turn to wsl.
when i open fox with gui in wsl, it occurs Segmentation fault ,now i wants to generate mesh and save it in wsl without gui, how should i do?

**Comment by Tom94** [comment]
Hi there, a couple of comments:
- Switching the operating system will not affect the mesh export function. I believe you're talking about bycould's video, in which case the suboptimal mesh came from the fact that they used `colmap` to estimate the lego digger's camera poses, rather than using the bundled `lego/transforms_train.json`.
- You can export the mesh without GUI using the Python bindings. In particular, you can use `testbed.compute_marching_cubes_mesh(resolution=256)` to get the vertices as numpy arrays and `testbed.compute_and_save_marching_cubes_mesh(filename, resolution=256)` to directly save it.

Cheers!

**Reply by Evilom** [reply]
thanks,i tested the bundled json, the result is fine, so how to generate a json like that? when i use my own data,the result also bad
![Snipaste_2022-02-23_00-27-28](https://user-images.githubusercontent.com/33251918/155176294-7472a4a2-183a-401f-80b2-74fbed8fb79e.png)

**Reply by Tom94** [reply]
Unfortunately, it's not going to get much better than this. __instant-ngp__ is not a dedicated meshing solution. There's no "regularization" that encourages the NeRF model to have plausible density values -- the loss only encourages good appearance.

If great meshing results are desired, I would recommend implementing RegNeRF or RefNeRF on top of our codebase, or, alternatively, using dedicated mesh optimization technique like [nvdiffrec](https://nvlabs.github.io/nvdiffrec/).
